akka-actor/src/main/scala/akka/util/Unsafe.java

```diff
@@ -23,8 +27,66 @@
             }
             if (found == null) throw new IllegalStateException("Can't find instance of sun.misc.Unsafe");
             else instance = found;
+            stringValueFieldOffset = instance.objectFieldOffset(String.class.getDeclaredField("value"));
+
+            // See Oracle section 1.5.3 at:
+            // https://docs.oracle.com/javase/8/docs/technotes/guides/versioning/spec/versioning2.html
+            final int[] version = Arrays.
+                stream(System.getProperty("java.specification.version").split("\\.")).
+                mapToInt(Integer::parseInt).
+                toArray();
+            javaVersion = version[0] == 1 ? version[1] : version[0];
         } catch (Throwable t) {
             throw new ExceptionInInitializerError(t);
         }
     }
+
+    public static void copyUSAsciiStrToBytes(String str, byte[] bytes) {
+        if (javaVersion > 8) {
+            final byte[] chars = (byte[]) instance.getObject(str, stringValueFieldOffset);
+            System.arraycopy(chars, 0, bytes, 0, chars.length);
+        } else {
+            final char[] chars = (char[]) instance.getObject(str, stringValueFieldOffset);
+            int i = 0;
+            while (i < chars.length) {
+                bytes[i] = (byte) chars[i++];
+            }
+        }
+    }
+
+    public static int fastHash(String str) {
+        long s0 = 391408;
+        long s1 = 601258;
+        int i = 0;
+
+        if (javaVersion > 8) {
+            final byte[] chars = (byte[]) instance.getObject(str, stringValueFieldOffset);
+            while (i < chars.length) {
+                long x = s0 ^ (long)chars[i++]; // Mix character into PRNG state
+                long y = s1;
+
+                // Xorshift128+ round
+                s0 = y;
+                x ^= x << 23;
+                y ^= y >>> 26;
+                x ^= x >>> 17;
+                s1 = x ^ y;
+            }
+        } else {
+            final char[] chars = (char[]) instance.getObject(str, stringValueFieldOffset);
+            while (i < chars.length) {
+                long x = s0 ^ (long)chars[i++]; // Mix character into PRNG state
+                long y = s1;
+
+                // Xorshift128+ round
+                s0 = y;
+                x ^= x << 23;
+                y ^= y >>> 26;
+                x ^= x >>> 17;
+                s1 = x ^ y;
+            }
+        }
```

Comment:
I guess it's not possible to DRY this without performance penalty.

* converting to `long[]` would mean another loop
* there is no common type for `byte[]` and `char[]` aside from `Object[]` and that would mean boxing

Reply:
It crossed my mind to use `System.arraycopy` treating both as `Object`s but `char` = 2 bytes and `byte` = 1 byte.